### PR TITLE
`TagsField` component

### PR DESCRIPTION
**Pull request by ggirotto**

### CODE REVIEW

**Review by matuella — COMMENTED** [review]
Waiting for changes discussed on a previous call. Added just a couple of minor comments.

File: lib/application/constants/dimensions.dart [hunk 1]
```diff
@@ -57,7 +57,19 @@ const double dragIndicatorWidth = 40;
...
 //
-const richTextFieldConstraints = BoxConstraints(maxHeight: 150, minHeight: 52);
+const richTextFieldConstraints = BoxConstraints(maxHeight: 148, minHeight: minTextFieldHeight);
+
+//
+// Tags Dropdown
+//
+const double tagsCloseButtonSize = 12;
+const double tagsSuggestionLoadingSize = 12;
+const double tagsSuggestionLoadingStrokeWidth = 2;
```
[the remark this answers is not in the document]
Comment (ggirotto) [reply lib/application/constants/dimensions.dart:75]: This is not a button. I'll rename the variable name to avoid this confusion.

File: lib/application/hooks/tags_controller_hook.dart [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_hooks/flutter_hooks.dart';
+import 'package:memo/application/widgets/theme/tags_field.dart';
+
+const useTagsController = _TagsControllerHookCreator();
```
Comment (matuella) [inline lib/application/hooks/tags_controller_hook.dart:5]: missing docs

File: test/application/widgets/theme/highlightable_text_test.dart [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import '../../../utils/widget_pump.dart';
+
+void main() {
+  const fakeText = 'A fake text';
+  const fakeTextStyle = TextStyle(color: Colors.white);
+  const fakeHighlight = 'fake';
+  const fakeHighlightStyle = TextStyle(color: Colors.red);
+
+  testWidgets('should not style highlight prefix', (tester) async {
```
Comment (matuella) [inline test/application/widgets/theme/highlightable_text_test.dart:13]: I think all of this tests could be merged into 2: `should use the default textStyle on anything other than the highlightable text` and `should use the highlightedStyle on highlighted text`.

File: test/application/widgets/theme/highlightable_text_test.dart [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:memo/application/widgets/theme/highlightable_text.dart';
+
+import '../../../utils/widget_pump.dart';
+
+void main() {
```
Comment (matuella) [inline test/application/widgets/theme/highlightable_text_test.dart:7]: Could add a test for scenarios where fakeHighlight is "empty" text, null or highlight not present in the actual text

Comment (matuella) [reply test/application/widgets/theme/highlightable_text_test.dart:7]: Could add a test for case-insentivity

File: test/application/widgets/theme/highlightable_text_test.dart [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    );
+    const expectedPrefixSpan = TextSpan(text: ' text', style: fakeTextStyle);
+
+    await pumpProviderScoped(tester, highlightableText);
+
+    final richText = tester.widget(find.byType(RichText)) as RichText;
+    final textSpan = richText.text as TextSpan;
+    final innerTextSpan = textSpan.children!.first as TextSpan;
+    expect(innerTextSpan.children!.last, expectedPrefixSpan);
+  });
```
Comment (matuella) [inline test/application/widgets/theme/highlightable_text_test.dart:62]:
```suggestion
    const highlightableText = HighlightableText(
      text: fakeText,
      textStyle: fakeTextStyle,
      highlighted: fakeHighlight,
      highlightedStyle: fakeHighlightStyle,
    );
    
     await pumpProviderScoped(tester, highlightableText);
     
    final richText = tester.widget(find.byType(RichText)) as RichText;
    final textSpan = (richText.text as TextSpan).children!.first as TextSpan;

  testWidgets('should use the default textStyle on anything other than the highlightable text', (tester) async {
    const expectedPrefixSpan = TextSpan(text: 'A ', style: fakeTextStyle);
    expect(textSpan.children![0], expectedPrefixSpan);
    
    const expectedSuffixSpan = TextSpan(text: ' text', style: fakeTextStyle);
    expect(textSpan.children![2], expectedSuffixSpan);
  });
  
  testWidgets('should use the highlightedStyle on highlighted text', (tester) async {
    const expectedHighlightedSpan = TextSpan(text: 'fake', style: fakeHighlightStyle);
    expect(textSpan.children![1], expectedHighlightedSpan);
  });
```

Comment (matuella) [reply test/application/widgets/theme/highlightable_text_test.dart:62]: ofc you won't have tester outside that scope, but you can work around that

File: lib/application/widgets/theme/tags_field.dart [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+import 'package:memo/application/constants/animations.dart' as anims;
+import 'package:memo/application/constants/dimensions.dart' as dimens;
+import 'package:memo/application/constants/images.dart' as images;
+import 'package:memo/application/constants/strings.dart' as strings;
+import 'package:memo/application/hooks/tags_controller_hook.dart';
+import 'package:memo/application/theme/theme_controller.dart';
+
+/// Provides the list of tags of the collection being updated.
+///
+/// These tags may be mutated by [TagsField], which the user may use to update the tags list.
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:16]: iMO these docs could be improved, they lack context and responsibilities

File: lib/application/widgets/theme/tags_field.dart [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+class TagsField extends HookConsumerWidget {
+  const TagsField({this.controller, this.maxTags = 5, this.errorText});
+
+  /// Controls the tags being edited.
+  ///
+  /// If null, this widget will create its own [TagsController].
+  final TagsController? controller;
+
+  /// {@template TagsDropdownField.maxTags}
+  /// Maximum number of tags allowed to be chosen.
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:35]:
```suggestion
  /// Maximum allowed tags to be selected simultaneously.
```

File: lib/application/widgets/theme/tags_field.dart [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+  ///
+  /// Inputs that would increase this amount will be ignored, i.e, not transformed in tags.
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:37]: Not sure exactly the meaning of this

File: lib/application/widgets/theme/tags_field.dart [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+  /// {@endtemplate}
+  final int maxTags;
+
+  /// {@macro CustomTextField.errorText}
+  final String? errorText;
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:45]: This widget kind-of looks too big, I think it could be split into one or multiple smaller widget(s). This would benefit in being more performant on updates, as well all of being specific on its responsibility, granting more "succinctness" to each widget/nested widget.

File: lib/application/widgets/theme/tags_field.dart [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    });
+
+    // Simulates `TextField.helperText` animating the label when the field is focused or not.
+    final helperTitle = AnimatedDefaultTextStyle(
+      style: isFocused.value ? textTheme.caption!.copyWith(color: theme.neutralSwatch.shade300) : textTheme.subtitle1!,
+      duration: anims.textFieldHelperTextDuration,
+      child: const Text(strings.addTags),
+    );
+
+    final tagsFieldContainer = Container(
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:99]: Like this, really looks like a child widget

File: lib/application/widgets/theme/tags_field.dart [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+  final void Function(String tag)? onTap;
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final theme = ref.watch(themeController);
+    final textTheme = Theme.of(context).textTheme;
+
+    return Material(
+      child: InkWell(
+        onTap: () => onTap?.call(tag),
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:234]:
```suggestion
        onTap: onTap != null ? onTap : null,
```

This will make sure all semantics and visuals are in order with inkwell

File: lib/application/widgets/theme/tags_field.dart [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+        focus?.requestFocus();
+      },
+    );
+
+    return Wrap(
+      spacing: context.rawSpacing(Spacing.small),
+      runSpacing: context.rawSpacing(Spacing.small),
+      children: [
+        ...tagsWidgets,
+        IntrinsicWidth(child: textField),
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:213]: IntrinsicWidth is really expensive on the UI, don't we have any alternative for this here?

File: lib/application/widgets/theme/tags_field.dart [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+  final TagsController controller;
+  final TextEditingController? fieldController;
+
+  /// Controls the current field focus.
+  final FocusNode? focus;
+
+  /// Called when the user indicates they are done inputting `tag`.
+  ///
+  /// Usually fired when the user press the "done" keyboard button.
+  final Function(String tag)? onSubmitted;
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:168]:
```suggestion
  final void Function(String tag)? onSubmitted;
```

File: lib/application/widgets/theme/tags_field.dart [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+  final TagsController controller;
+  final TextEditingController? fieldController;
+
+  /// Controls the current field focus.
+  final FocusNode? focus;
+
+  /// Called when the user indicates they are done inputting `tag`.
+  ///
+  /// Usually fired when the user press the "done" keyboard button.
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:167]: Is there any other possibility other than tapping "done"?

File: lib/application/widgets/theme/tags_field.dart [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+          ).withAllPadding(context, Spacing.xxSmall),
+        ),
+      ),
+    );
+  }
+}
+
+/// Formats the text from [_TagsTextField] `TextField`.
+class _TagFieldFormatter extends TextInputFormatter {
+  static const _maxTagLength = 15;
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:256]: shouldn't this be a value provided by some validation class? This looks to me like a leaked business definition

File: lib/application/widgets/theme/tags_field.dart [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+  @override
+  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
+    if (newValue.text.length > _maxTagLength) {
+      return TextEditingValue(text: oldValue.text, selection: oldValue.selection);
+    }
+
+    // Restrict the content to only accept alphanumeric space, underscore and comma characters.
+    final alphanumRegex = RegExp(r'^[a-zA-Z0-9_ ,]*$');
```
Comment (matuella) [inline lib/application/widgets/theme/tags_field.dart:265]: ditto: shouldn't this be a value provided by some validation class? This looks to me like a leaked business definition

File: test/application/widgets/theme/tags_field_test.dart [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+      controller.tags = [firstFakeTag, secondFakeTag];
+      await tester.pumpAndSettle();
+
+      expect(find.text(firstFakeTag), findsOneWidget);
+      expect(find.text(secondFakeTag), findsOneWidget);
+    });
+  });
+}
+
+/// Pumps [field] optionally adding [tag].
```
Comment (matuella) [inline test/application/widgets/theme/tags_field_test.dart:167]:
```suggestion
/// Pumps [field] and calls [_addTag] using [tag].
```

File: test/application/widgets/theme/tags_field_test.dart [hunk 18]
```diff
@@ -0,0 +1,181 @@
+import 'package:flutter/material.dart';
```
Comment (matuella) [inline test/application/widgets/theme/tags_field_test.dart:1]: Really nice looking test coverage, well done

File: test/application/widgets/theme/tags_field_test.dart [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+import 'package:flutter_test/flutter_test.dart';
+import 'package:memo/application/constants/strings.dart' as string;
+import 'package:memo/application/theme/theme_controller.dart';
+import 'package:memo/application/widgets/theme/tags_field.dart';
+
+import '../../../utils/widget_pump.dart';
+
+void main() {
```
Comment (matuella) [inline test/application/widgets/theme/tags_field_test.dart:9]: What about a test that deals with "min" length (or not accepting empty strings)?

**Review by matuella — CHANGES_REQUESTED** [review]
Suggestion for future PRs: this one could've been split into several smaller PRs, and they would be completely fine/independent, but changes are looking good.

File: lib/application/constants/dimensions.dart [hunk 20]
```diff
@@ -57,7 +57,17 @@ const double dragIndicatorWidth = 40;
...
 //
 // Rich Text Field
 //
-const richTextFieldConstraints = BoxConstraints(maxHeight: 148, minHeight: 52);
+const richTextFieldConstraints = BoxConstraints(maxHeight: 148, minHeight: minTextFieldHeight);
+
+//
+// Tags Dropdown
+//
+const double tagsRemoveIconSize = 12;
```
[the remark this answers is not in the document]
Comment (ggirotto) [reply lib/application/constants/dimensions.dart:73]: Unfortunately not. This is a very specific scenario that needs a very tiny icon size. Take a look at the SS:

<details>
<summary>Screenshot</summary>

CleanShot 2021-11-20 at 10 26 21@2x

</details>

File: test/application/widgets/theme/highlightable_text_test.dart (hunk 4 above)
Comment (ggirotto) [reply test/application/widgets/theme/highlightable_text_test.dart:7]: Nice suggestions

File: test/application/widgets/theme/highlightable_text_test.dart (hunk 5 above)
Comment (ggirotto) [reply test/application/widgets/theme/highlightable_text_test.dart:62]: Added a function to to that. Let me know what do you think.

File: lib/application/widgets/theme/tags_field.dart (hunk 6 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:16]: I've improved the docs but also renamed from `TagsController` to `TagsEditingController` to follow the same naming pattern as `TextEditingController` and create the analogy that it has the same purpose but with tags. Let me know what do you think.

File: lib/application/widgets/theme/tags_field.dart (hunk 8 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:37]: Removed the second sentence because it didn't added any relevant information

File: lib/application/widgets/theme/tags_field.dart (hunk 12 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:213]: Unfortunately not. This is one of those scenarios where the `Intrinsic` Widgets must be used, otherwise it would be possible to place the `TextField` used to input the tags right next to the tags itself, because the `TextField` would stretch to fit the entire width

File: lib/application/widgets/theme/tags_field.dart (hunk 13 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:168]: Can't we have a lint for that? I'll take a look

File: lib/application/widgets/theme/tags_field.dart (hunk 14 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:167]: Updated the docs, the "done" button is the only possibility here

File: lib/application/widgets/theme/tags_field.dart (hunk 15 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:256]: Definitely. I'll address this issue in https://github.com/olmps/memo/pull/237 because there I've added the validators.

File: lib/application/widgets/theme/tags_field.dart (hunk 9 above)
Comment (ggirotto) [reply lib/application/widgets/theme/tags_field.dart:45]: I moved the `tagsFieldContainer` to a separate Widget which improved a little bit the main Widget size. I tried some different approaches trying to shrink the `useEffect` size but the problem is that the 3 controllers (1. tags controller 2. text field controller and 3. focus controller) are very coupled because the act of selecting a tag impact in all of them. Let me know what do you think.

**Review by matuella — APPROVED** [review]
